lib/unleash/repo.ex

```diff
@@ -1,124 +1,128 @@
+  elapsed, refreshing both our local ETS cache and the backup state file if the
+  flag state has diverged.
+
+  The following configuration values are used:
+
+  Config.features_period(): polling interval - default 15 seconds
+  Config.retries(): number of time the call to refresh values is allowed to retry - default -1 (0)
+  """
+
   use GenServer
   require Logger
 
   alias Unleash.Config
   alias Unleash.Features
+  alias Unleash.Cache
 
   def init(%Features{} = features) do
+    Cache.init(features.features)
     {:ok, features}
   end
 
   def init(_) do
+    Cache.init()
     {:ok, %Features{}}
   end
 
   def start_link(state) do
     {:ok, pid} = GenServer.start_link(__MODULE__, state, name: Unleash.Repo)
 
-    unless Config.test? do
+    unless Config.test?() do
       initialize()
     end
 
     {:ok, pid}
   end
 
   def get_feature(name) do
-    GenServer.call(Unleash.Repo, {:get_feature, name}, 100)
+    Cache.get_feature(name)
   end
 
   def get_all_feature_names do
-    GenServer.call(Unleash.Repo, {:get_all_feature_names}, 100)
-  end
-
-  def handle_call({:get_feature, name}, _from, state) do
-    feature = Features.get_feature(state, name)
-
-    {:reply, feature, state}
-  end
-
-  def handle_call({:get_all_feature_names}, _from, state) do
-    names = Features.get_all_feature_names(state)
-    {:reply, names, state}
+    Cache.get_all_feature_names()
   end
 
-  def handle_info({:initialize, etag, retries}, state) do
+  def handle_info({:initialize, etag, retries}, _state) do
     if retries > 0 or retries <= -1 do
+      cached_features = %Features{features: Cache.get_features()}
+
       {etag, response} =
         case Unleash.Config.client().features(etag) do
-          :cached -> {etag, state}
+          :cached -> {etag, cached_features}
           x -> x
         end
 
-      features =
+      remote_features =
         case response do
           {:error, _} ->
-            state
-            |> read_state()
+            cached_features
+            |> read_file_state()
             |> schedule_features(etag, retries - 1)
 
           f ->
             schedule_features(f, etag)
         end
 
-      if features === state do
-        {:noreply, state}
+      if remote_features === cached_features do
+        {:noreply, nil}
       else
-        write_state(features)
+        Cache.upsert_features(remote_features.features)
+        write_file_state(remote_features)
+        {:noreply, nil}
       end
     else
       Logger.debug(fn ->
         "Retries === 0, disabling polling"
       end)
 
-      {:noreply, state}
+      {:noreply, nil}
     end
   end
 
   # https://github.com/appcues/mojito/issues/57
   # Work around for messages received from Mojito after we've passed over the timeout
   # threshold.
-  def handle_info({:mojito_response, _ref, _message}, state) do
-    {:noreply, state}
+  def handle_info({:mojito_response, _ref, _message}, _state) do
+    {:noreply, nil}
```

Comment:
you initialise state to `%Feature{}` but change it here to `nil` instead of keeping the original state

Reply:
ah if `%Features{}` is used then use that as the state across the board and don't overwrite it with `nil`

Reply:
I changed it to be assigned to `[]` on init and not to re-assign it to nil in the `handle_info` handler